### PR TITLE
[TensorRate/Filter] Skip inference when QoS events arrive

### PR DESCRIPTION
**Pull request by dongju-chae**
This patch implements a simple QoS mechanism which skips
inference when QoS events arrive from tensor_rate.      

The below shows test results using tflite object detection example

- Without throttling (original pipeline, eventually saturated 10 FPS due to tensor filter's performance)
```
...
  ! queue leaky=2 max-size-buffers=2 ! videoscale ! video/x-raw,width=640,height=480 ! tensor_converter
  ! tensor_transform mode=arithmetic option=typecast:float32,add:-127.5,div:127.5
  ! tensor_filter framework=tensorflow-lite model=...
...
```
without throttling

- With throttling (`tensor_rate` added, and no `max-size-buffer` setting, we want to limit framerate to 5 FPS)
```
...
  ! queue ! videoscale ! video/x-raw,width=640,height=480 ! tensor_converter
  ! tensor_transform mode=arithmetic option=typecast:float32,add:-127.5,div:127.5
  ! tensor_filter framework=tensorflow-lite model=...
  ! tensor_rate framerate=5/1 throttle=true
...
```

with throttling

In short, even without limiting queue's max buffers, we can adjust a frame rate to 5 FPS, and effectively reduces CPU consumption.

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

### CODE REVIEW

File: gst/nnstreamer/tensor_filter/tensor_filter.c [hunk 1]
```diff
@@ -456,6 +461,29 @@ gst_tensor_filter_transform (GstBaseTransform * trans,
       GST_STR_NULL (prop->model_files[0]));
   allocate_in_invoke = gst_tensor_filter_allocate_in_invoke (priv);
 
+  /* skip input data when throttling delay is set */
+  if (priv->throttling_delay != 0) {
```
Comment (kparichay) [inline gst/nnstreamer/tensor_filter/tensor_filter.c:465]: As per our offline discussion, this would be best applied to tensor_converter and tensor_src*

Comment (dongju-chae) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:465]: Yes, sure. I will make next PRs for that.

File: gst/nnstreamer/tensor_filter/tensor_filter.c [hunk 2]
```diff
@@ -999,6 +1050,48 @@ gst_tensor_filter_sink_event (GstBaseTransform * trans, GstEvent * event)
...
+      if (type == GST_QOS_TYPE_THROTTLE && diff > 0) {
+        priv->throttling_delay = diff;
+        gst_event_unref (event);
+
+        /* enable the average latency profiling */
+        g_object_set (self, "latency", 1, NULL);
+        return TRUE;
+      }
+
+      return gst_pad_push_event (trans->sinkpad, event);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_filter/tensor_filter.c:1085]: default evt handle in bast-transform?

Comment (dongju-chae) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:1085]: Updated.

File: gst/nnstreamer/tensor_filter/tensor_filter.c (hunk 1 above)
Comment (myungjoo) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:479]: What if...
```
t.c (doing 100FPS) --> T -+-> T-F model A (may do 10 FPS)
                          +-> T-F model B (may do 200 FPS)
```
Do you still want to handle overflow QoS events at t.c?

Comment (dongju-chae) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:479]: Yes, I have to. But, this PR is targeting to compare its latency and throttling delay from `tensor_rate`, as a prototype.
I will make another PR to consider multiple QoS events from downstream elements.

**Review by myungjoo — COMMENTED** [review]
Before merging, please describe the QoS policy of tensor_filter as a GST Element Manual entry (at the top section)

**Comment by dongju-chae** [comment]
@myungjoo Sure. Its manual will be updated soon.

**Review by myungjoo — APPROVED** [review]
LGTM.

To reviewers: this is a huge update on the core. Please review it carefully.

**Review by helloahn — APPROVED** [review]
LGTM

**Comment by myungjoo** [comment]
PTAL: @jaeyun-jung @kparichay

File: gst/nnstreamer/tensor_filter/tensor_filter.c [hunk 3]
```diff
@@ -457,6 +487,44 @@ gst_tensor_filter_transform (GstBaseTransform * trans,
...
+    if (GST_CLOCK_TIME_IS_VALID (prev_ts)) {
+      GstClockTimeDiff diff = curr_ts - prev_ts;
+      GstClockTimeDiff delay;
+
+      priv->throttling_accum += diff;
+
+      /* check whether the average latency is longer than throttling delay */
+      delay = MAX (priv->prop.latency * 1000, priv->throttling_delay);
+
+      if (priv->throttling_accum < delay) {
```
Comment (kparichay) [inline gst/nnstreamer/tensor_filter/tensor_filter.c:508]: should this be `throttling_accum > delay`? throttling_accum indicates the time between consecutive frames. If this frame exceeds the allowed, then QoS is send. 
With the current code, if `curr_ts == prev_ts`, `throttling_accum` goes to 0, and a QoS event is sent upstream with positive diff as `delay` (from documentation - A positive diff value indicates how late the buffer with timestamp was.). However, in this case, the buffer is not at all late.

File: gst/nnstreamer/tensor_filter/tensor_filter.c (hunk 1 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:493]: @dongju-chae I wonder if the images in the PR description have been updated or not? Because the fps of tensor_converter is 36FPS. However, with this change, we expect it to also be 5fps.
Please correct me if I understood wrongly.

File: gst/nnstreamer/tensor_filter/tensor_filter.c (hunk 3 above)
Comment (dongju-chae) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:508]: > should this be throttling_accum > delay? throttling_accum indicates the time between consecutive frames. If this frame exceeds the allowed, then QoS is send.

No. `throttling_accum` is being used to drop frames. So, until it reaches to `delay`, all incoming frames should be dropped.

Comment (kparichay) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:508]: Understood, I misunderstood buffer timestamp.

**Review by kparichay — APPROVED** [review]
LGTM

File: gst/nnstreamer/tensor_filter/tensor_filter.c (hunk 1 above)
Comment (dongju-chae) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:493]: Yep, it seems like a glitch. It keeps slightly changing over time.
